mainnet_contract/contracts/Subnet.sol

```diff
@@ -0,0 +1,214 @@
+    if (!masters[msg.sender]) revert("Masters Only");
+    _;
+  }
+
+  constructor(address[] memory initial_validator_set, int threshold, bytes memory genesis_header) public {
+    require(initial_validator_set.length > 0, "Validator Empty");
+    bytes32 genesis_header_hash = keccak256(genesis_header);
+    (bytes32 ph, int n) = HeaderReader.getParentHashAndNumber(genesis_header);
+    header_tree[genesis_header_hash] = Header({
+      hash: genesis_header_hash,
+      number: n,
+      round_num: 0, 
+      parent_hash: ph,
+      finalized: true,
+      mainnet_num: block.number,
+      src: genesis_header
+    });
+    validator_sets[0] = Validators({
+      set: initial_validator_set,
+      threshold: threshold
+    });
+    for (uint i = 0; i < initial_validator_set.length; i++) {
+      lookup[initial_validator_set[i]] = true;
+    }
+    masters[msg.sender] = true;
+    latest_finalized_block = genesis_header_hash;
+  }
+
+  function isMaster(address master) public view returns (bool) {
+    return masters[master];
+  }
+
+  function addMaster(address master) public onlyMasters {
+    masters[master] = true;
+  }
+
+  function removeMaster(address master) public onlyMasters {
+    masters[master] = false;
+  }
+
+  function reviseValidatorSet(address[] memory new_validator_set, int threshold, int subnet_block_height) public onlyMasters  {
+    require(new_validator_set.length > 0, "Validator Empty");
+    require(threshold > 0, "Invalid Threshold");
+    require(subnet_block_height >= current_validator_set_pointer, "Modify Past Validator");
+    validator_sets[subnet_block_height] = Validators({
+      set: new_validator_set,
+      threshold: threshold
+    });
+  }
+
+  function receiveHeader(bytes memory header) public onlyMasters { 
+    (bytes32 parent_hash, int number, uint64 round_number, bytes32 signHash, bytes[] memory sigs) = HeaderReader.getValidationParams(header);
+    require(number > 0, "Repeated Genesis");
+    require(header_tree[parent_hash].hash != 0, "Parent Missing");
+    require(header_tree[parent_hash].number + 1 == number, "Invalid N");
+    require(header_tree[parent_hash].round_num < round_number, "Invalid RN");
+    bytes32 block_hash = keccak256(header);
+    if (header_tree[block_hash].number > 0) 
+      revert("Repeated Header");
+    if (validator_sets[number].set.length > 0) {
+      for (uint i = 0; i < validator_sets[current_validator_set_pointer].set.length; i++) {
+        lookup[validator_sets[current_validator_set_pointer].set[i]] = false;
+      }
+      for (uint i = 0; i < validator_sets[number].set.length; i++) {
+        lookup[validator_sets[number].set[i]] = true;
+      }
+      current_validator_set_pointer = number;
+    }
+
+    int unique_counter = 0;
+    address[] memory signer_list = new address[](sigs.length);
+    for (uint i = 0; i < sigs.length; i++) {
+      address signer = recoverSigner(signHash, sigs[i]);
+      if (lookup[signer] != true) {
+        continue;
```

Comment:
We can enforce the header contain no repeated signatures, so here we can revert. And the logic here will also be simpler